### PR TITLE
multi: enforce conservative minimum for user-chosen CLTV deltas

### LINKED ISSUE #4468: cltv_expiry for invoice shouldn't be less than 9
### Background

Tried making a loop in transaction that kept failing because the incoming HTLC kept getting cancelled. On detailed examination, it turns out my setting `bitcoin.timelockdelta` to `6` made the generated invoice have a `cltv_expiry` of `6` which is too low as the BOLT11 spec requires the `final_cltv_expiry` to be at least `9`.

### Your environment

* v0.10.3-beta
* Linux 4.9.0-11-amd64 #1 SMP Debian 4.9.189-3+deb9u2 (2019-11-11) x86_64 GNU/Linux
* bitcoin-core v0.20.0

### Steps to reproduce

Set configuration parameter `bitcoin.timelockdelta` to `6`; restart lnd and then create an invoice for any amount. Now try to pay the invoice and lnd will cancel the htlc.

### Expected behaviour

lnd should generate invoices with the value for `cltv_expiry` being no less than `9`.

### Actual behaviour

lnd will generate invoices with the `cltv_expiry` set to whatever value is configured for `bitcoin.timelockdelta` even if it is less than `9`.

### PR DESCRIPTION
**Pull request by cfromknecht**
This PR modifies LND's validation of user-chosen CLTV values by enforcing a more conservative value of 18 blocks for both forwarding and final CLTV deltas. Currently LND uses a default value of 40 blocks for both values, but allows the user to choose a lower value if desired. Forwarded CLTV deltas are required to be at least 4 blocks, while final CLTV deltas have no minimum. 

The spec is currently upgrading its recommended minimum from 9 to 18,  so this PR clamps both to be at least 18 blocks. In doing so, we ensure that LND rejects potentially insecure CLTV deltas to give itself ample time to address on-chain handling scenarios. 

Along the way, we further isolate `migration_01_to_11` which currently imports `zpay32` directly in order to decode invoices. This is a preparatory yet critical step since any further changes to `zpay32` could impact the correctness of the migration, so the existing deserialization code has been committed into a subpackage to ensure no further changes are made. This will be important in the future once the default _assumed_ final CLTV value is increased in the spec, as the migration might assume the wrong final CLTV delta if none is decoded.

Fixes #4468.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -73,7 +73,7 @@ const (
 
 	// minTimeLockDelta is the minimum timelock we require for incoming
 	// HTLCs on our channels.
-	minTimeLockDelta = 4
+	minTimeLockDelta = routing.MinCLTVDelta
```
Comment (Roasbeef) [inline config.go:76]: Hmm, perhaps we should use a distinct constant for this value? This sets the min _CLTV delta_ allowed (when verifying against a users config values), while in this PR the main aim is to modify the min _final CTLV delta_ which governs the acceptable block delta for HTLC presented to the final hop in a route.

File: routing/router.go [hunk 2]
```diff
@@ -46,6 +46,27 @@ const (
...
+	// than the sender, the receiver will always see the received HTLCs as
+	// invalid due to their timelock not meeting the required delta.
+	//
+	// We skirt this by always setting an explicit CLTV delta when creating
+	// invoices. This allows LND nodes to freely update the minimum without
+	// creating incompatibilities during the upgrade process. For some time
+	// LND has used an explicit default final CLTV delta of 40 blocks for
+	// bitcoin (160 for litecoin), though we now clamp the lower end of this
+	// range for user-chosen deltas to 18 blocks to be conservative.
+	MinCLTVDelta = 18
```
Comment (Roasbeef) [inline routing/router.go:69]: See comment above, we may want to prepend this with "final". On the other-hand, we could just enforce the min configurable routing CLTV delta from the min invoice/payment final CLTV delta.

File: config.go (hunk 1 above)
Comment (cfromknecht) [reply config.go:76]: > while in this PR the main aim is to modify the min final CTLV delta which governs the acceptable block delta for HTLC presented to the final hop in a route.

no, the goal is to clamp both forwarded and final CLTV deltas, which is the [change being made to the spec](https://github.com/lightningnetwork/lightning-rfc/pull/785). both will have a minimum recommended value of 18, there really isn't a reason to distinguish the two

File: routing/router.go (hunk 2 above)
Comment (cfromknecht) [reply routing/router.go:69]: if the routing CLTV is just going to be set to the final CLTV delta, why differentiate?

Comment (Roasbeef) [reply routing/router.go:69]: Good point, works as is.

**Pull request merged**